backend/src/main/controller/PromotionController.ts

```diff
@@ -0,0 +1,188 @@
+import { NextFunction, Request, Response } from 'express';
+import { PromotionRepository } from '../repository/PromotionRepository';
+import { getCustomRepository, SelectQueryBuilder } from 'typeorm';
+import { Promotion } from '../entity/Promotion';
+import { UserRepository } from '../repository/UserRepository';
+import { Discount } from '../entity/Discount';
+import {
+  PromotionDTO,
+  PromotionValidation,
+} from '../validation/PromotionValidation';
+import {
+  PromotionQueryDTO,
+  PromotionQueryValidation,
+} from '../validation/PromotionQueryValidation';
+import { IdValidation } from '../validation/IdValidation';
+
+/* eslint-disable  @typescript-eslint/no-explicit-any */
+export class PromotionController {
+  /**
+   * Retrieves all promotions and their discounts
+   * * First we need to validate the query params and cast that into a PromotionQueryDTO
+   * * Then we apply the query options into the query builder depending on which properties are present
+   * * Now we execute the query builder and return its results
+   */
+  getAllPromotions = async (
+    request: Request,
+    response: Response,
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      let queryBuilder = getCustomRepository(PromotionRepository)
+        .createQueryBuilder('promotion')
+        .innerJoinAndSelect('promotion.discount', 'discount');
```

Comment:
For this endpoint I could not use a repository, unfortunately. But using the queryBuilder isn't that complicated. The reaosn why we need query builder is that if we want to filter by discount type, then we need to join. 

We cannot just do 
```
getCustomRepository(PromotionRepository).find({where: discount {type: '%'}, relations: ['discount']}
````